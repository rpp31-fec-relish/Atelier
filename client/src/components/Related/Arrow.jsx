import React from 'react';

const Arrow = ({ direction, handleClick, glyph, currentImageIndex }) => {
  if (currentImageIndex === 0 || currentImageIndex === 3) {
    glyph = '';
  }
  return (
    <div className={`slide-arrow ${direction}`} onClick={ handleClick }>{ glyph }
    </div>
  )
}

export default Arrow;