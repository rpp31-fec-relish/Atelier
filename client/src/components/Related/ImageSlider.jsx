import React from 'react';

const ImageSlider = ({ url }) => {
  const styles = {
    backgroundImage: `url(${url})`,
    backgroundSize: 'cover',
    backgroundPosition: 'center'
  };

  return (
    <div className="carousel-item" style={styles}></div>
  )
}

export default ImageSlider;