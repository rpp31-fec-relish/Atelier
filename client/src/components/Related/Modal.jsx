import React from 'react';

export default class Modal extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      allStyles: []
    }

    this.onClose = this.onClose.bind(this);
  }

  onClose(e) {
    this.props.onClose && this.props.onClose(e);
  }

  render() {
    if (!this.props.show) {
      return null;
    }

    return (
      <table id="RP-modal">
        <div className="modal-title">COMPARING</div>
        <tr>
          <th>Current Product Name</th>
          <th> </th>
          <th>Compared Product Name</th>
        </tr>
        <tr>
          <td>&#10003;</td>
          <td>GMO and Pesticide-free</td>
          <td>&#10003;</td>
        </tr>
        <tr>
          <td>&#10003;</td>
          <td>Made with 100% Genetic Modification</td>
          <td>&#10003;</td>
        </tr>
        <tr>
          <td>&#10003;</td>
          <td>This is made up</td>
          <td>&#10003;</td>
        </tr>
        <tr>
          <td>&#10003;</td>
          <td>It doesn't matter</td>
          <td></td>
        </tr>
        <tr>
          <td>&#10003;</td>
          <td>Feature description</td>
          <td></td>
        </tr>
        <tr>
          <td></td>
          <td>Uses React Hooks and Redux</td>
          <td>&#10003;</td>
        </tr>
        <tr>
          <td></td>
          <td>Angular</td>
          <td>&#10003;</td>
        </tr>
        <tr>
          <td></td>
          <td>Some other product comparison metric</td>
          <td>&#10003;</td>
        </tr>
        <div>
          <div id="toggle-modal" onClick={(e) => {this.onClose(e);}}>&#9746;</div>
        </div>
      </table>
    )
  }
}