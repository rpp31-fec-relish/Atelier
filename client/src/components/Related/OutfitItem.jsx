import React from 'react';
import Stars from '../../../../node_modules/react-stars-display/';

const OutfitItem = (props) => {
  let {id, name, category, price, image, assignImage, outfits, addToOutfit, changeCurrentProduct, productRating, handleRemove} = props;

  return (
    <td id="OutfitItem">
      <div id={id} className="Outfit-x-button" onClick={(e) => {handleRemove(id)}}>&#9746;</div>
      <input id="Outfit-image" type="image" alt="image" src={assignImage(image)} onClick={() => changeCurrentProduct(id)}></input>
      <div>{category}</div>
      <div>{name}</div>
      <div>{price}</div>
      <div id="Outfit-rating">
        <Stars stars={productRating ? productRating : 0}/>
      </div>
    </td>
  )
};

export default OutfitItem;