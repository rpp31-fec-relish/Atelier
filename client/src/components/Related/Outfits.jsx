import React from 'react';

const Outfits = (props) => {
  if (props.outfits.length === 0) {
    return '';
  } else {
    return (
      <section id="OutfitsWidget">
        <div id="OutfitItem">
          <p>
            <input type="image" alt="image"></input>
            <h6>Category</h6>
            <h6>Title</h6>
            <h6>Price</h6>
            <h6>Rating</h6>
          </p>
        </div>
        <div id="OutfitItem">
          <p>
            <input type="image" alt="image"></input>
            <h6>Category</h6>
            <h6>Title</h6>
            <h6>Price</h6>
            <h6>Rating</h6>
          </p>
        </div>
        <div id="OutfitItem">
          <p>
            <input type="image" alt="image"></input>
            <h6>Category</h6>
            <h6>Title</h6>
            <h6>Price</h6>
            <h6>Rating</h6>
          </p>
        </div>
        <div id="OutfitItem">
          <p>
            <input type="image" alt="image"></input>
            <h6>Category</h6>
            <h6>Title</h6>
            <h6>Price</h6>
            <h6>Rating</h6>
          </p>
        </div>
      </section>
    )
  }
}

export default Outfits;