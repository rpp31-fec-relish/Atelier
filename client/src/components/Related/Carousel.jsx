import React from 'react';
import Arrow from './Arrow.jsx';
import RelatedProduct from './RelatedProduct.jsx';
import OutfitItem from './OutfitItem.jsx';

class Carousel extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      relatedIndex: 0,
      outfitIndex: 0
    };

    this.previousSlide = this.previousSlide.bind(this);
    this.nextSlide = this.nextSlide.bind(this);
    this.indexSet = this.indexSet.bind(this);
  }

  previousSlide(widget) {
    if (widget === 'related') {
      if (this.state.relatedIndex !== 0) {
        const index = this.state.relatedIndex - 1;
        this.setState({
          relatedIndex: index
        })
      }
    } else if (widget === 'outfits') {
      if (this.state.outfitIndex !== 0) {
        const index = this.state.outfitIndex - 1;
        this.setState({
          outfitIndex: index
        })
      }
    }
  }

  nextSlide(widget) {
    if (widget === 'related') {
      if (this.state.relatedIndex !== 4) {
        const index = this.state.relatedIndex + 1;
        this.setState({
          relatedIndex: index
        });
      }
    } else if (widget === 'outfits') {
      if (this.state.outfitIndex !== 3) {
        const index = this.state.outfitIndex + 1;
        this.setState({
          outfitIndex: index
        });
      }
    }
  }

  indexSet(index, widget) {
    if (widget === 'related') {
      let maxIndex = this.state.relatedIndex + index;
      return maxIndex;
    } else if (widget === 'outfits') {
      let maxIndex = this.state.outfitIndex + index;
      return maxIndex;
    }
  }

  render() {
    if (this.props.widget === 'related') {
      return (
        <div className='RP-carousel'>
          <div className='RP-arrow'>
            <Arrow
              direction='left'
              handleClick={ this.previousSlide }
              glyph='&#9664;'
              relatedIndex={this.state.relatedIndex}
              max={this.props.data.length}
              widget={this.props.widget}
            />
          </div>
          <table>
            <tbody>
              <tr>
                {this.props.data.slice(this.state.relatedIndex, this.indexSet(4, this.props.widget)).map((item) => <RelatedProduct
                  key={'relatedProduct_' + item.id}
                  id={item.id}
                  name={item.name}
                  category={item.category}
                  price={item.price}
                  image={item.image}
                  assignImage={this.props.assignImage}
                  changeCurrentProduct={this.props.changeCurrentProduct}
                  showModal={this.props.showModal}
                />)}
              </tr>
            </tbody>
          </table>
          <div className='RP-arrow'>
            <Arrow
            direction='right'
            handleClick={ this.nextSlide }
            glyph='&#9654;'
            relatedIndex={this.state.relatedIndex}
            max={this.props.data.length}
            widget={this.props.widget}
            />
          </div>
        </div>
      )
    } else if (this.props.widget === 'outfits') {
      return (
        <div className='Outfits-carousel'>
          <div className='RP-arrow'>
            <Arrow
              direction='left'
              handleClick={ this.previousSlide }
              glyph='&#9664;'
              outfitIndex={this.state.outfitIndex}
              max={this.props.outfitData.length}
              widget={this.props.widget}
            />
          </div>
          <table>
            <tbody>
              <tr>
                <td className='Outfit-add-remove'>
                  <div onClick={(e) => {this.props.handleClick(e)}}>{this.props.handleTextChange()}</div>
                </td>
                {this.props.outfitData.slice(this.state.outfitIndex, this.indexSet(3, this.props.widget)).map((outfit) => <OutfitItem
                  key={'outfit_' + outfit.id}
                  id={outfit.id}
                  name={outfit.name}
                  category={outfit.category}
                  price={outfit.price}
                  image={outfit.image}
                  removeOutfit={this.props.removeOutfit}
                  assignImage={this.props.assignImage}
                  handleClick={this.props.handleClick}
                  addToOutfit={this.props.addToOutfit}
                  changeCurrentProduct={this.props.changeCurrentProduct}
                  productRating={this.props.productRating}
                />)}
              </tr>
            </tbody>
          </table>
          <div className='RP-arrow'>
            <Arrow
              direction='right'
              handleClick={ this.nextSlide }
              glyph='&#9654;'
              outfitIndex={this.state.outfitIndex}
              max={this.props.outfitData.length}
              widget={this.props.widget}
            />
          </div>
          {/* <div className="Outfit-add-remove">
            <div onClick={(e) => {this.props.handleClick(e)}}>{this.props.handleTextChange()}</div>
          </div>
          {this.props.outfitData.slice(this.state.outfitIndex, this.indexSet(3, this.props.widget)).map((outfit) => <OutfitItem
            key={'outfit_' + outfit.id}
            id={outfit.id}
            name={outfit.name}
            category={outfit.category}
            price={outfit.price}
            image={outfit.image}
            removeOutfit={this.props.removeOutfit}
            assignImage={this.props.assignImage}
            handleClick={this.props.handleClick}
            addToOutfit={this.props.addToOutfit}
            changeCurrentProduct={this.props.changeCurrentProduct}/>)}
          <Arrow
            direction='right'
            handleClick={ this.nextSlide }
            glyph='&#9654;'
            outfitIndex={this.state.outfitIndex}
            max={this.props.outfitData.length}
            widget={this.props.widget}/>
        </div> */}
        </div>
      )
    }
  }
}

export default Carousel;