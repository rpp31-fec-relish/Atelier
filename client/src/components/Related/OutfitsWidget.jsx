import React from 'react';
import OutfitItem from './OutfitItem.jsx';
import helperFunctions from './../../helperFunctions.js';

class OutfitsWidget extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      outfitsId: [],
      outfitsData: []
    };

    this.handleTextChange = this.handleTextChange.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.getOutfitData = this.getOutfitData.bind(this);
  }

  handleTextChange() {
    if (this.props.outfits.includes(this.props.currentProduct)) {
      return 'REMOVE FROM OUTFIT';
    } else {
      return 'ADD TO OUTFIT';
    }
  }

  getOutfitData() {
    let newData = [];

    helperFunctions.getProductById(this.props.currentProduct)
    .then(productData => {
      let data = {
        id: productData.id,
        name: productData.name,
        category: productData.category,
        price: productData.default_price,
        features: productData.features,
        image: null
      }
      this.setState({ outfitsId: [...this.state.outfitsId, productData.id] })
      newData.push(data);
      return data;
    })
    .then(results =>  {
      helperFunctions.getProductStylesById(results.id)
        .then(styles => {
          results.image = styles[0].photos;
          return results;
        })
        .then(relevantData => {
          this.setState({outfitsData: [...this.state.outfitsData, relevantData]});
        })
        .catch(err => console.error(err));
    })
    .catch(err => console.error(err));
  }

  handleClick(e) {
    e.preventDefault();
    if (this.state.outfitsId.includes(this.props.currentProduct)) {
      let newOutfitsData = [...this.state.outfitsData];
      newOutfitsData.forEach(outfit => {
        if (outfit.id === this.props.currentProduct) {
          var removeIndex = newOutfitsData.map(item => item.id).indexOf(this.props.currentProduct);
          ~removeIndex && newOutfitsData.splice(removeIndex, 1);
        }
      })
      this.setState({outfitsData: newOutfitsData})
    } else {
      this.getOutfitData();
      this.props.addToOutfit(this.props.currentProduct);
    }
  }

  render() {
    return (
      <section id="OutfitsWidget">
        <div id="OutfitItem">
          <div id="addToOutfit">
            <a href='#add-outfit' id="OutfitText" onClick={(e) => {
              this.handleClick(e);
            }}>+ {this.handleTextChange()}</a>
          </div>
        </div>
        {this.state.outfitsData.map((outfit) => <OutfitItem key={'outfit_' + outfit.id} id={outfit.id} name={outfit.name} category={outfit.category} price={outfit.price} image={outfit.image} assignImage={this.props.assignImage} outfits={this.props.outfits}/>)}
      </section>
    )
  }
}

export default OutfitsWidget;