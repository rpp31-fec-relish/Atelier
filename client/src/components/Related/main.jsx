import React from 'react';
import RelatedProductsWidget from './RelatedProductsWidget.jsx';
import OutfitsWidget from './OutfitsWidget.jsx';
import helperFunctions from '../../helperFunctions.js';
import Modal from './Modal.jsx';

class Related extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      show: false,
      relatedProductFeatures: [],
      relatedProductsId: [],
      relatedProductsData: [],
      singleRelatedProductFeature: [],
      allStyles: [],
      outfitsData: []
    }

    this.assignImage = this.assignImage.bind(this);
    this.showModal = this.showModal.bind(this);
    this.updateOutfitsData = this.updateOutfitsData.bind(this);
  }

  componentDidMount() {
    let newData = [];
    let relatedFeatures = [];

    helperFunctions.getRelatedProductsById(this.props.currentProduct)
      .then(products => {
        for (var i = 0; i < products.length; i++) {
          let data = {
            id: products[i].id,
            name: products[i].name,
            category: products[i].category,
            price: products[i].default_price,
            image: null
          }
          let relatedData = {
            id: products[i].id,
            name: products[i].name,
            features: products[i].features
          }
          if (!this.state.relatedProductsId.includes(products[i].id)) {
            this.setState({ relatedProductsId: [...this.state.relatedProductsId, products[i].id] })
            newData.push(data);
            relatedFeatures.push(relatedData);
          }
        }

        this.setState({relatedProductFeatures: relatedFeatures});
        return newData;
      })
      .then(results => {
        results.forEach(product => {
          helperFunctions.getProductStylesById(product.id)
            .then(productStyle => {
              if (productStyle.length > 0) {
                for (var i = 0; i < results.length; i++) {
                  if (results[i].id === product.id) {
                    results[i].image = productStyle[0].photos;
                  }
                }
              } else {
                console.error('No product styles');
              }
              return results;
            })
            .then(results => {
              let noDuplicates = [];
              results.forEach(item => {
                if (item.id !== parseInt(this.props.currentProduct)) {
                  noDuplicates.push(item);
                }
              })
              return noDuplicates;
            })
            .then(relevantData => {
              this.setState({relatedProductsData: relevantData});
            })
            .catch(err => console.error(err));
        })
      })
      .catch(err => console.error(err));
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevProps.currentProduct !== this.props.currentProduct) {
      this.setState({ relatedProductsId: [] })
      this.componentDidMount();
    }
    if (prevProps.outfits.length + 1 === this.props.outfits.length && this.props.outfits.includes(this.props.currentProduct) && !prevProps.outfits.includes(this.props.currentProduct)) {
      this.updateOutfitsData('add');
    } else if (prevProps.outfits.length - 1 === this.props.outfits.length && !this.props.outfits.includes(this.props.currentProduct) && prevProps.outfits.includes(this.props.currentProduct)) {
      this.updateOutfitsData('remove');
    }
  }

  updateOutfitsData(action) {
    if (action === 'add' || !action) {
      let CPD = this.props.currentProductData;
      let removeDuplicates = this.state.outfitsData.map(item => item.id).indexOf(CPD.id);
      if (removeDuplicates > -1) {
        let newOutfitsData = [...this.state.outfitsData];
        newOutfitsData.splice(removeDuplicates, 1);
        this.setState({outfitsData: newOutfitsData});
        this.handleTextChange();
      } else {
        let data = {
          id: CPD.id,
          name: CPD.name,
          category: CPD.category,
          price: CPD.default_price,
          image: null
        };

        helperFunctions.getProductStylesById(CPD.id)
          .then(styles => {
            data.image = styles[0].photos;
            return data;
          })
          .then(relevantData => {
            this.setState({outfitsData: [...this.state.outfitsData, relevantData]});
          })
          .catch(err => console.error(err));
      }
    } else if (action === 'remove') {
      let currentId = this.props.currentProduct;
      let newOutfitsData = [...this.state.outfitsData];
      newOutfitsData.forEach(outfit => {
        if (outfit.id === parseInt(currentId)) {
          var removeIndex = newOutfitsData.map(item => item.id).indexOf(parseInt(currentId));
          ~removeIndex && newOutfitsData.splice(removeIndex, 1);
        }
      })
      this.setState({outfitsData: newOutfitsData});
    // for the outfit item's 'x' button, action = the id
    } else {
      let newOutfitsData = [...this.state.outfitsData];
      newOutfitsData.forEach(outfit => {
        if (outfit.id === action) {
          var removeIndex = newOutfitsData.map(item => item.id).indexOf(action);
          ~removeIndex && newOutfitsData.splice(removeIndex, 1);
        }
      })
      this.setState({outfitsData: newOutfitsData});
    }
  }

  assignImage(imageArray) {
    const PlaceholderPhoto = './images/missingImage.svg';

    if (imageArray) {
      if (imageArray[0].thumbnail_url) {
        return imageArray[0].thumbnail_url;
      } else {
        return PlaceholderPhoto;
      }
    }
    return PlaceholderPhoto;
  }

  showModal(e) {
    if (e.target.id) {
      let comparisonFeature = [];
      let uniqueStyles = [];
      let relatedFeatures = this.state.relatedProductFeatures;
      let currentFeatures = this.props.currentProductData.features;

      for (var i = 0; i < relatedFeatures.length; i++) {
        if (relatedFeatures[i].id === parseInt(e.target.id)) {
          let newFeatures = {
            feature: relatedFeatures[i].features,
            name: relatedFeatures[i].name
          }
          comparisonFeature.push(newFeatures);
        }
      }

      for (var i = 0; i < currentFeatures.length; i++) {
        if (currentFeatures[i].value == null) {
          currentFeatures[i].value = ''
        }
        if (currentFeatures[i].feature == null) {
          currentFeatures[i].feature = ''
        }
        let currentCharacteristic = `${currentFeatures[i].value} ${currentFeatures[i].feature}`.replace(/([A-Z])/g, ' $1').trim();
        let modalEntry = {
          current: true,
          characteristic: [currentCharacteristic],
          compare: false
        }
        uniqueStyles.push(modalEntry);
      }

      if (comparisonFeature[0]) {
        comparisonFeature[0].feature.forEach(feature => {
          if (feature.value == null) {
            feature.value = ''
          }
          if (feature.feature == null) {
            feature.feature = ''
          }
          let comparisonFeature = `${feature.value} ${feature.feature}`.replace(/([A-Z])/g, ' $1').trim();
          for (var i = 0; i < uniqueStyles.length; i++) {
            if (uniqueStyles[i].characteristic == comparisonFeature) {
              uniqueStyles[i].compare = true;
              break;
            } else if (i+1 === uniqueStyles.length) {
              let modalEntry = {
                current: false,
                characteristic: [comparisonFeature],
                compare: true
              }
              uniqueStyles.push(modalEntry);
            }
          }
        })
      }

      for (var i = 0; i < uniqueStyles.length; i++) {
        uniqueStyles[i].key = i;
      }

      this.setState({ allStyles: uniqueStyles });
      this.setState({ singleRelatedProductFeature: comparisonFeature })
    }
    this.setState({ show: !this.state.show });
  };

  render() {
    return (
      <div>
        <Modal
          onClose={this.showModal}
          show={this.state.show}
          currentProductData={this.props.currentProductData}
          singleRelatedProductFeature={this.state.singleRelatedProductFeature}
          allStyles={this.state.allStyles}></Modal>
        <div id="RelatedProductsAndOutfits">
          <h4 className="RP-titles">RELATED PRODUCTS</h4>
          <RelatedProductsWidget
            currentProduct={this.props.currentProduct}
            assignImage={this.assignImage}
            changeCurrentProduct={this.props.changeCurrentProduct}
            showModal={this.showModal}
            setRelatedProductFeatures={this.setRelatedProductFeatures}
            relatedProductsData={this.state.relatedProductsData}/>
          <h4 className="RP-titles">YOUR OUTFITS</h4>
          <OutfitsWidget
            currentProduct={this.props.currentProduct}
            outfits={this.props.outfits}
            assignImage={this.assignImage}
            addToOutfit={this.props.addToOutfit}
            changeCurrentProduct={this.props.changeCurrentProduct}
            showModal={this.showModal}
            currentProductData={this.props.currentProductData}
            productRating={this.props.productRating}
            outfitsData={this.state.outfitsData}
            updateOutfitsData={this.updateOutfitsData}/>
        </div>
      </div>
    );
  }
}

export default Related;