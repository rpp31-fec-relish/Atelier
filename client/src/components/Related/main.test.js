import Related from './main.jsx';

import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

// describe - it - expect
describe('Related', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Related />);
    expect(wrapper).toMatchSnapshot();
  });
});