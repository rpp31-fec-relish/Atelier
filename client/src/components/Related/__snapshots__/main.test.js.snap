// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Related renders correctly 1`] = `
<div>
  <ul
    id="related"
    style={
      Object {
        "border": "1px solid black",
      }
    }
  >
    <RelatedProducts />
    <Outfits />
  </ul>
</div>
`;
